src/init.rs

```diff
@@ -59,3 +52,97 @@ pub fn init(shell_name: &str) {
+   command is part of a pipeline. To avoid only timing the last part of a pipeline,
+   we only start the timer if no timer has been started since the last prompt draw,
+   tracked by the variable PREEXEC_READY. Similarly, only draw timing info if
+   STARSHIP_START_TIME is defined, in case preexec was interrupted.
+
+   Finally, note that `eval` will evaluate this thing as a single line, so things
+   like comments (#) should be avoided and semicolons should be used on most
+   statements (but not after then/else).
+*/
+
+const BASH_INIT: &str = r##"
+starship_preexec() {
+    if [ "$PREEXEC_READY" = "true" ]; then
+        PREEXEC_READY=false;
+        STARSHIP_START_TIME=$(date +%s);
+    fi
+};
+starship_precmd() {
+    STATUS=$?;
+    if [[ $STARSHIP_START_TIME ]]; then
+        STARSHIP_END_TIME=$(date +%s);
+        STARSHIP_ELAPSED=$((STARSHIP_END_TIME - STARSHIP_START_TIME));
+        PS1="$(starship prompt --status=$STATUS --elapsed=$STARSHIP_ELAPSED)";
+        unset STARSHIP_START_TIME;
+    else
+        PS1="$(starship prompt --status=$STATUS)";
+    fi;
+    PREEXEC_READY=true;
+};
+if [[ $preexec_functions ]]; then
+    preexec_functions+=(starship_preexec);
+    precmd_functions+=(starship_precmd);
+    STARSHIP_START_TIME=$(date +%s);
+fi;
+dbg_trap="$(trap -p DEBUG | cut -d' ' -f3 | tr -d \')";
+if [[ -z $dbg_trap || "$dbg_trap" = "starship_preexec" ]]; then
+    trap starship_preexec DEBUG;
+    PROMPT_COMMAND=starship_precmd;
+    STARSHIP_START_TIME=$(date +%s);
+else
+    PROMPT_COMMAND=starship_precmd;
+fi
+"##;
+//TODO: How to warn the user?
+
+/* For zsh: preexec_functions and precmd_functions provide preexec/precmd in a
+   way that lets us avoid clobbering them.
+
+   Zsh quirk: preexec() is only fired if a command is actually run (unlike in
+   bash, where spamming empty commands still triggers DEBUG). This means a user
+   spamming ENTER at an empty command line will see increasing runtime (since
+   preexec never actually fires to reset the start time).
+
+   To fix this, only pass the time if STARSHIP_START_TIME is defined, and unset
+   it after passing the time, so that we only measure actual commands.
+*/
+
+const ZSH_INIT: &str = r##"
+starship_precmd() {
+    if [[ $STARSHIP_START_TIME ]]; then
+        STARSHIP_END_TIME="$(date +%s)";
+        ELAPSED=$((STARSHIP_END_TIME - STARSHIP_START_TIME));
+        PROMPT="$(starship prompt --status=$? --elapsed=$ELAPSED)";
+        unset STARSHIP_START_TIME;
+    else
+        PROMPT="$(starship prompt --status=$?)";
+    fi
+};
+starship_preexec(){
+    STARSHIP_START_TIME="$(date +%s)"
+};
+precmd_functions+=(starship_precmd);
+preexec_functions+=(starship_preexec);
+STARSHIP_START_TIME="$(date +%s)";
+"##;
+
+/* Fish setup is simple because they give us CMD_DURATION. Hooray! */
+const FISH_INIT: &str = r##"
+function fish_prompt;
+    set -l elapsed (math --scale=0 "$CMD_DURATION / 1000");
```

Comment:
I would suggest adding:
```fish
set -l command_duration "$CMD_DURATION$cmd_duration"
```

Between fish 2.7 and 3.0, the casing for the variable was changed.